language-hask-source/src/Language/Hask/Parsed/AST.hs

```diff
@@ -0,0 +1,575 @@
+    --
+    -- > module $1
+  deriving stock Show
+
+-- | Parsed import declaration.
+--
+-- > import $importQualified $importName $importAlias $importHiding ($importList)
+data ImportDeclaration = ImportDeclaration{
+    importQualified :: Bool             -- ^ Presence of @qualified@ keyword.
+  , importName      :: ModuleName
+  , importAlias     :: Maybe ModuleName -- ^ Presense of @as $_@ syntax.
+  , importHiding    :: Bool             -- ^ Presence of @hiding@ keyword.
+  , importList      :: [Import]
+  } deriving stock Show
+
+-- | Parsed import.
+data Import
+  = VariableImport VariableName
+    -- ^ Import of a simple variable.
+  | TypeImport ConstructorName (Maybe TypeImportList)
+    -- ^ Import of a type.
+    --
+    -- > $1
+    --
+    -- or
+    --
+    -- > $1 ($2)
+    --
+    -- See [Parsed type import/export lists](#ParsedTypeImportLists).
+  deriving stock Show
+
+-- | Parsed import/export list of a type. You may be interested in
+-- [Parsed type import/export lists](#ParsedTypeImportLists).
+data TypeImportList
+  = TypeImportAll
+    -- ^ Import/export of all members in scope.
+    --
+    -- > (..)
+  | TypeImportList [Either VariableName ConstructorName]
+    -- ^ Import/export of explicitly named members.
+    --
+    -- > ($1)
+  deriving stock Show
+
+-- $ParsedTypeImportLists
+-- == Parsed type import/export lists #ParsedTypeImportLists#
+-- 'TypeImportList's contain both class and datatype imports/exports, because
+-- we can't yet distinguish those properly - that's up to the renamer.
+
+-------------------------------------------------------------------------------
+-- | Parsed top-level declaration.
+data Declaration
+  = TypeSynonym Type Type
+    -- ^ Type synonym declaration.
+    --
+    -- > type $1 = $2
+  | Datatype Type DatatypeBody (Maybe Deriving)
+    -- ^ Datatype declaration.
+    --
+    -- > data $1 = $_ | ... $3
+    --
+    -- or
+    --
+    -- > newtype $1 = $2 $3
+    --
+    -- See [Types in declaration heads](#TypesInDeclarationHeads).
+  | ClassDeclaration Class
+  | InstanceDeclaration Instance
+  | Defaults [Type]
+    -- ^ "default" declaration.
+    --
+    -- > default ($1)
+  | Binding Binding
+  deriving stock Show
+
+-- | Parsed datatype body.
+data DatatypeBody = Newtype NewtypeConstructor | Data [DataConstructor]
+  deriving stock Show
+
+-- | Parsed deriving clause.
+--
+-- > deriving $_
+--
+-- or
+--
+-- > deriving ($_, ...)
+type Deriving = Either (Qualified ConstructorName) [Qualified ConstructorName]
+
+-- | Parsed "newtype" constructor.
+--
+-- > $newtypeConstructorName { $newtypeConstructorLabel :: $newtypeConstructorField }
+data NewtypeConstructor = NewtypeConstructor{
+    newtypeConstructorName  :: ConstructorName
+  , newtypeConstructorLabel :: Maybe VariableName
+  , newtypeConstructorField :: Type
+  } deriving stock Show
+
+-- | Parsed "data" constructor.
+--
+-- > $dataConstructorName $dataConstructorFields
+data DataConstructor = DataConstructor{
```

Comment:
I don't know if I fully follow what the `Left` in the `Either` represents here. I would want something like:
```haskell
data DT v1 ... vn where
  DC1 :: forall v'1 ... v'n . Ty1 -> ... -> Tyl -> DT v1 ... vn
```
to translate to something like:
```haskell
data DataConstructor = DataConstructor
  { dataConstructorName :: ConstructorName
  , univVars           :: [Vars]
  , quantifiedVars  :: [Vars]
  , constructorArgs :: [Type]
  }
```
Here is an example of how I imagine this parsing:
```haskell
data E a b where
  L :: forall b a . a -> E a b
  R :: forall b a . b -> E a b
```
would translate to:
```haskell
DataConstructor "L" [Var "a", Var "b"] [Var "b", Var "a"] [TyVar (Var "a")]
```
Am I thinking too much about GADTs here?

Reply:
See comment below - according to those principles, in case of support for `GADTs`, we could take e.g.:
```hs
data GADTConstructor = GADTConstructor ConstructorName Type
```

Reply:
Just to be clear, the Left case in `dataConstructorFields` would be for concrete types like:
```
data Foo = Foo Int Bool
```
And the Right case is for GADTs?

Reply:
`Right` is for named fields in record syntax - should I put it under synonym for clarity?
Oh, wait, I missed bangs in `Left` case! I wonder whether I should put them into `Type` and error later or just fix that.